addons/account/tests/test_account_payment_items.py

```diff
@@ -0,0 +1,181 @@
+        cls.late_bill.invoice_date_due = '2023-03-01'
+        for line in cls.late_bill.line_ids:
+            line.date_maturity = cls.late_bill.invoice_date_due
+
+        cls.discount_bill = cls.env['account.move'].create({
+            'move_type': 'in_invoice',
+            'journal_id': journal.id,
+            'partner_id': cls.private_partner.id,
+            'date': '2023-03-15',
+            'invoice_date': '2023-03-01',
+            'invoice_line_ids': [(0, 0, {
+                'product_id': cls.product_a.id,
+                'quantity': 1,
+                'name': 'product_a',
+                'discount': 10.00,
+                'price_unit': 100,
+                'tax_ids': [],
+            })]
+        })
+        cls.discount_bill.action_post()
+        cls.discount_bill.invoice_date_due = '2023-03-25'
+        for line in cls.discount_bill.line_ids:
+            line.write({
+            'discount_date': '2023-03-20',
+            'date_maturity': cls.discount_bill.invoice_date_due,
+        })
+
+        cls.late_discount_bill = cls.env['account.move'].create({
+            'move_type': 'in_invoice',
+            'journal_id': journal.id,
+            'partner_id': cls.private_partner.id,
+            'date': '2023-03-15',
+            'invoice_date': '2023-03-01',
+            'invoice_line_ids': [(0, 0, {
+                'product_id': cls.product_a.id,
+                'quantity': 1,
+                'name': 'product_a',
+                'discount': 10.00,
+                'price_unit': 100,
+                'tax_ids': [],
+            })]
+        })
+        cls.late_discount_bill.action_post()
+        cls.late_discount_bill.invoice_date_due = '2023-04-20'
+        for line in cls.late_discount_bill.line_ids:
+            line.write({
+                'discount_date': '2023-03-10',
+                'date_maturity': cls.late_discount_bill.invoice_date_due,
+                    })
+
+    @freeze_time("2023-03-15")
+    def test_payment_date(self):
+        # Testing that the payment_date on the lines of a waiting for a
+        # payment move without discount is the date_maturity
+        self.assertEqual(str(self.bill.line_ids[0].payment_date), '2023-03-20')
+        # Testing that the payment_date on the lines of a late move without
+        # discount is the date_maturity
+        self.assertEqual(str(self.late_bill.line_ids[0].payment_date), '2023-03-01')
+        # Testing that the payment_date on the lines of a waiting for a payment
+        # move with a discount date that is not passed yet is the discount_date
+        self.assertEqual(str(self.discount_bill.line_ids[0].payment_date), '2023-03-20')
+        # Testing that the payment_date on the lines of a waiting for a payment
+        # move with a discount date that is already passed is the date_maturity
+        self.assertEqual(str(self.late_discount_bill.line_ids[0].payment_date), '2023-04-20')
+
+    def test_search_payment_date(self):
```

Comment:
This is already a nice test :+1: , but what I believe would make it more readable is something like
```python
# bill:               due 2023-03-20
# late_bill:          due 2023-03-01
# discount_bill:      due 2023-03-25 discount 2023-03-20
# late_discount_bill: due 2023-04-01 discount 2023-03-10 
for today, search, expected in [
    ('2023-03-05', '2023-03-01', late_bill),
    ('2023-03-05', '2023-03-30', bill + late_bill + discount_bill + late_discount_bill),
    ...
]:
    with freeze_time(today):
        self.assertEqual(self.env['account.move.line'].search([
            ('payment_date', '=', search),
            ('partner_id', '=', self.private_partner_id.id),
        ]).move_id, expected)
```

Instead of the comments, a helper function that takes those as arguments would be even better
```python
bill = create_bill(due='2023-03-20')
late_bill = create_bill(due='2023-03-01')
discount_bill = create_bill(due='2023-03-25', discount='2023-03-20')
late_discount_bill = create_bill(due='2023-04-01', discount='2023-03-10')
```


This way all the information is at the same place and I think the comments become useless.

_____

In general, always [Keep it DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)